Move executor pod construction to a separate class.

mccheah opened the pull request:
This is the first of several measures to make KubernetesClusterSchedulerBackend feasible to test. Requires #445 but only for convenience and not semantically speaking.

The idea is to start breaking down the functionality of KubernetesClusterSchedulerBackend into multiple individually unit-test friendly units. The logic that builds the executor pod structure was by far the single longest method that could be isolated with relative ease.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:
```diff
@@ -0,0 +1,296 @@
+    }.getOrElse(containerWithExecutorLimitCores)
+    val withMaybeShuffleConfigPod = shuffleServiceConfig.map { config =>
+      config.shuffleDirs.foldLeft(executorPod) { (builder, dir) =>
+        new PodBuilder(builder)
```
Ah we lost the indentation here, I'll fix that.

ash211 replied:
looks fixed now

foxish commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:
```diff
@@ -0,0 +1,294 @@
+      ConfigurationUtils.parsePrefixedKeyValuePairs(
+          sparkConf,
+          KUBERNETES_NODE_SELECTOR_PREFIX,
+          "node-selector")
```
"node selector" for consistency?

mccheah replied:
That would be a break in the configuration I think. Aside from that, `SparkConf` keys have never had spaces in them.

ash211 replied:
I think that last string is only used for log output -- `node selector` seems like it would be fine

foxish commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala:
```diff
@@ -234,18 +234,28 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
         "-DsimpleDriverConf=simpleDriverConfValue" +
             " -Ddriverconfwithspaces='driver conf with spaces value'")
-    sparkConf.set("spark.files", driverJvmOptionsFile.getAbsolutePath)
+    sparkConf.set(SparkLauncher.EXECUTOR_EXTRA_JAVA_OPTIONS,
```
This is a separate change from the refactor correct?

foxish reviewed (commented):
Refactor LGTM

mccheah commented:
I believe the diff is corrupted and I have to fix it in git.

mccheah commented:
Actually I just need to rebase against branch-2.2-kubernetes and make the diff that way.

mccheah committed:
Move executor pod construction to a separate class.

This is the first of several measures to make
KubernetesClusterSchedulerBackend feasible to test.

ash211 commented on resource-managers/kubernetes/README.md:
```diff
@@ -14,14 +14,11 @@ important matters to keep in mind when developing this feature.
-If this is the first time you compile the Kubernetes core implementation module, run the following command to install the dependencies and compile:
-
-    build/mvn install -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
-
```
please don't revert this change -- it went in in a PR and somehow your PRs keep coming close to reverting it.. ?

mccheah commented on resource-managers/kubernetes/README.md:
```diff
@@ -14,14 +14,11 @@ important matters to keep in mind when developing this feature.
+To build Spark with Kubernetes support, use the `kubernetes` profile when invoking Maven. For example, to simply compile
+the Kubernetes core implementation module along with its dependencies:
 
     build/mvn compile -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
```
Believe this change is unintentional

mccheah committed:
Revert change to README

mccheah commented:
@foxish rebase complete.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:
```diff
@@ -0,0 +1,294 @@
+import org.apache.spark.deploy.kubernetes.submit.{InitContainerUtil, MountSmallFilesBootstrap}
+import org.apache.spark.util.Utils
+
+// Strictly an extension of KubernetesClusterSchedulerBakcne that is factored out for testing.
```
typo: `KubernetesClusterSchedulerBakcne`

ash211 replied:
by extension, you mean the method in this trait is the same as a method in `KubernetesClusterSchedulerBackend` ?  That implies to me it should do multiple inheritance.

I'm not sure the "strictly an extension" language makes sense -- maybe instead say that it's only used in the scheduler backend?

mccheah replied:
Extension meaning a plugin, or functionality that is pretty much only used in the scheduler backend.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:
```diff
@@ -0,0 +1,294 @@
+
+  import ExecutorPodFactoryImpl._
+
+  private val EXECUTOR_ID_COUNTER = new AtomicLong(0L)
```
where is this used?  I think it should be deleted because KubernetesClusterSchedulerBackend.scala already has one

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/NodeAffinityExecutorPodModifier.scala:
```diff
@@ -0,0 +1,84 @@
+import org.apache.spark.deploy.kubernetes.constants.ANNOTATION_EXECUTOR_NODE_AFFINITY
+import org.apache.spark.internal.Logging
+
+// Strictly an extension of ExecutorPodFactory but extracted out for testing.
```
not sure this comment adds much -- it's good practice to have smaller more modular pieces anyway for understanding, regardless of testing purposes

mccheah committed:
Address comments.

ash211 commented:
@mccheah conflicts on the GiB -> MiB conversion fix since you moved that elsewhere

mccheah committed:
Merge remote-tracking branch 'origin/branch-2.2-kubernetes' into separate-executor-pod-construction

mccheah committed:
Resolve merge conflicts.

Move MiB change to ExecutorPodFactory.

mccheah commented:
rerun integration tests please

ash211 committed:
Merge branch 'branch-2.2-kubernetes' into separate-executor-pod-construction

ash211 commented:
Rerun unit tests please

mccheah commented:
@aash @foxish good to merge this?

ash211 commented:
test coverage is unchanged (this is an internal refactor) and enables more granular testing in followup PRs
